### ISSUE TITLE
Question on how to integrate the GUI and CLI together for the iP

### ISSUE THREAD

**Issue by mingyang143** [nus-cs2103-AY2425S1/forum#253]
Context:
My iP now supports operations from both the CLI and a GUI. Also, when I type the "bye" command in my CLI, while terminating the CLI, I can close the GUI window automatically as well. To implement both CLI and GUI simultaneously, I utilized The Java Thread class to create new threads.

Problem:
However, I am unable to do the same the other way around. When I close my GUI, I am unable to terminate the CLI automatically. I have a hunch that the thread running the GUI (thread 1) is unable to close the thread running the CLI (thread 2) as thread 2 has an active scanner that is actively waiting for a new line. I have tried ways to interrupt the scanner but I am still unable to solve the issue.

Looking for some advice on how to solve the problem of automatically terminating the CLI when closing the GUI window. Thank you for reading.

**Comment by damithc** [comment]
@mingyang143 Technical issue aside, I'm curious about the rationale for having both UIs at the same time. Also, is there a way for users to choose one only (I expect most users would prefer one or the other but not both at the same time)?

**Comment by mingyang143** [comment]
Hi prof,

Thank you for your quick response.

My rationale is more about learning and exploring the possibility of implementing both UIs, rather than practical reasons.

At the moment, the user cannot choose between the UIs, but I believe I can implement this functionality with relative ease.

**Comment by damithc** [comment]
> My rationale is more about learning and exploring the possibility of implementing both UIs, rather than practical reasons.

Understood. Let's wait to see if others have any thoughts on this.

**Comment by yadobler** [comment]
Hello! Great point. I decided to keep the CLI, by separating the bot into its own object, and use another `Ui` interface that determines the contract between the bot and the I/O. Then I used one with CLI and one with GUI and then have the launcher class decide which UI to attach to the bot.

Choice is done with a switch (`-c` for CLI, omit for GUI) when running. If needed, this can be placed in a `.ink` or `.desktop` shortcut file.

I'll copy paste my comment [from another issue](https://github.com/nus-cs2103-AY2425S1/forum/issues/244#issuecomment-2337727375):

---

> After understanding what I need to interface with the GUI, I've come to the conclusion that my current codebase needs quite a bit of refactoring.

If it's to do with the main loop depending on `Scanner` blocking until input is available, you can get around with:

<details> <summary> what i did basically </summary>

1. extracting that `hasNext()` into an abstract method in `Ui` interface
2. have 2 different classes implement `Ui`, with the *cli* version's `hasNext` `running scanner.hasNext()`
3. **Cry** for awhile realising you need to implement some concurrency stream implementation for GUI (**This is important** because you know you should not exert too much effort into this but you can't help it)
4. Make your UiGui implementation of `Ui` hold a buffer of input and outputs. Basically instead of `System.out.println` going to `STDOUT` and scanner reading from `STDIN`, implement a _blocking_ function to busy wait until the input buffer / output buffer is not empty anymore and return true (or return false when your bot has ended, indicating a _closed stream_)
5. Run your bot in a new `Thread` so that it will concurrently pull from `UiGui`-IN and push to `UiGui`-OUT
6. Run another loop in your `MainWindow` (probably in the init method) class to run a new `Thread` to repeatably check if the Output buffer is not empty, and then _print_ it out as the bot's dialogbox

</details>

**tl;dr**
1. encapsulate Ui to be an interface doing the STDIN/OUT/ERR. Cli implementation can use normal System.out and System.in
2. Run your bot in a new thread so that it can _busy wait_ for _Input_
3. Run a seperate thread in your GUI application to _busy wait_ for _Output_ and **signal the main app to add new `DialogBox`s when output availabe (ie bot replies)**

Some caveats:
- <details> <summary> JavaFX updates must happen on the JavaFX line. </summary> Use `Platform.runLater(updateMethod)` to run a `updateMethod()` that does the `Ui.StdOut->DialogBox->Add to scene` </details>
- <details> <summary> I had to use a flag to add a guard clause in my update method. </summary> Idk why even though I only execute the `runLater` when the busyloop continues, it still gets stuck afterwards. _I think `Platform.runLater` might be executed once, so the new thread only needs to update the flag and not call it again. I will need to try this._ </details>
- Your program will be a bit slower on startup.

Why I did it this way:
- My bot says hello before the user inputs, so I don't want to have edge cases to handle with beginning and replying only when user inputs. Very messy.
- I am programmer. of course I will waste hours automating something that can be hacked away in minutes
- My CLI still works. [Run my .jar](https://github.com/yadobler/ip/releases/tag/Level-10) with `-c` flag to use CLI instead of GUI. This means my `Text-ui-test` still works, so **regression testing still works**
- I dont have to touch my main bot class anymore. All logic remains same. If tomorrow @damithc says the bot needs to run on brain with the input coming from my thought and output injected to my eyes, i just need a new `UiBrain implements Ui` class and add a new method to my `Launcher` static class to create the appropriate Objects for the bot.

---

- [My `UpdateMethod` method](https://github.com/yadobler/ip/blob/38cc4e82b0e247d5e5db9a2c44d369cb5299718f/src/main/java/yappingbot/ui/gui/MainWindow.java#L63C3-L76C1)
- [The new thread inside the GUI application](https://github.com/yadobler/ip/blob/38cc4e82b0e247d5e5db9a2c44d369cb5299718f/src/main/java/yappingbot/ui/gui/MainWindow.java#L44-L49)
- [The new thread running my bot, including disabling the GUI once the bot exits and returns](https://github.com/yadobler/ip/blob/38cc4e82b0e247d5e5db9a2c44d369cb5299718f/src/main/java/yappingbot/ui/gui/MainGuiApplication.java#L45-L50)

---

**If it is not, then ignore this.**
